From 4880a41e17b5cfa057e3f161fb502b0a8378ac52 Mon Sep 17 00:00:00 2001
From: Tim Smith <tim.smith@cloud.com>
Date: Mon, 16 Sep 2024 09:38:19 +0100
Subject: [PATCH 15/30] Re-revert 9e7787c

Re-revert commit 9e7787c after a merge from before it was reverted put
some of it back.

Signed-off-by: Tim Smith <tim.smith@cloud.com>
---
 drivers/tapdisk-nbdserver.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/drivers/tapdisk-nbdserver.c b/drivers/tapdisk-nbdserver.c
index 17c14238..34e9831b 100644
--- a/drivers/tapdisk-nbdserver.c
+++ b/drivers/tapdisk-nbdserver.c
@@ -841,8 +841,6 @@ tapdisk_nbdserver_free_client(td_nbdserver_client_t *client)
 	if (likely(!tapdisk_nbdserver_reqs_pending(client))) {
 		list_del(&client->clientlist);
 		tapdisk_nbdserver_reqs_free(client);
-		if (client->client_fd > 0)
-			close(client->client_fd);
 		free(client);
 	} else
 		client->dead = true;
@@ -1126,6 +1124,7 @@ tapdisk_nbdserver_newclient_fd_old(td_nbdserver_t *server, int new_fd)
 	if (tapdisk_nbdserver_enable_client(client) < 0) {
 		ERR("Error enabling client");
 		tapdisk_nbdserver_free_client(client);
+		close(new_fd);
 	}
 }
 
@@ -1153,6 +1152,7 @@ tapdisk_nbdserver_newclient_fd_new_fixed(td_nbdserver_t *server, int new_fd)
 	if(tapdisk_nbdserver_new_protocol_handshake(client, new_fd) != 0) {
 		ERR("Error handshaking new client connection");
 		tapdisk_nbdserver_free_client(client);
+		close(new_fd);
 		return;
 	}
 
@@ -1160,6 +1160,7 @@ tapdisk_nbdserver_newclient_fd_new_fixed(td_nbdserver_t *server, int new_fd)
 	if (tapdisk_nbdserver_enable_client(client) < 0) {
 		ERR("Error enabling client");
 		tapdisk_nbdserver_free_client(client);
+		close(new_fd);
 	}
 }
 
@@ -1282,8 +1283,11 @@ tapdisk_nbdserver_clientcb(event_id_t id, char mode, void *data)
 
 		break;
 	case TAPDISK_NBD_CMD_DISC:
-		INFO("Received close message. Free client");
+		INFO("Received close message. Sending reconnect header");
 		tapdisk_nbdserver_free_client(client);
+		INFO("About to send initial connection message");
+		tapdisk_nbdserver_newclient_fd(server, fd);
+		INFO("Sent initial connection message");
 		return;
 	case TAPDISK_NBD_CMD_BLOCK_STATUS:
 	{
